### ISSUE TITLE
cp: Running multiple instance of mc to same location

### ISSUE THREAD

**Issue by limbo127** [minio/mc#2057]
Hello, 
mc create a temporary name during copying  , -miniopart. 
I think this file must be timestamped, as s3cmd do, -miniopart<timestamp> because  : 
If I start multiple mc with same copy command,  destination copy is concatenated 

Regards, 
Nicolas

**Comment by vadmeste** [comment]
@limbo127, timestamp is not enough actually, because mc supports resume copy, if it finds any timestamped minio part file , it will continue to write the remaining data in that file. Then, two parallel mc copy will just write on the same part file.

@harshavardhana, we probably need some locking capabilities, I saw minio/pkg/lock package, but it supports only opening file for reading purposes, can't we add write support in the package?

**Comment by harshavardhana** [comment]
> @harshavardhana, we probably need some locking capabilities, I saw minio/pkg/lock package, but it supports only opening file for reading purposes, can't we add write support in the package?

We don't need that.. it is highly unlikely that same filename is copied by multiple `mc` and written to same location.  We are like coreutils `cp` i am not sure this is handled by coreutils `cp` command.

**Comment by limbo127** [comment]
cp does not concatenate file when same copy occur in parallel
regards,
Nicolas

2017-04-04 2:04 GMT+02:00 Harshavardhana <notifications@github.com>:

**Comment by vadmeste** [comment]
> cp does not concatenate file when same copy occur in parallel

Actually, two parallel copies can result a corrupted file. You can generate a 1 Gb file and try to copy two files to the same destination `for i in $(seq 1 2) ; do ( cp 1gb.file newfile & )  ; done ;`

Below is my test in Ubuntu 16.04, you can notice that the size of the destination file is not the same 

```
$ for i in $(seq 1 2) ; do ( cp 1gb.file newfile & )  ; done ; ls -l newfile
-rw-rw-r-- 1 vadmeste vadmeste 0 Apr  4 11:33 newfile

$ for i in $(seq 1 2) ; do ( cp 1gb.file newfile & )  ; done ; ls -l newfile
-rw-rw-r-- 1 vadmeste vadmeste 1073741824 Apr  4 11:33 newfile

$ for i in $(seq 1 2) ; do ( cp 1gb.bin newfile & )  ; done ; ls -l newfile
-rw-rw-r-- 1 vadmeste vadmeste 926408704 Apr  4 11:42 newfile
```

I also think that doing a generic solution for all platforms (linux, windows, ..) supported by mc is a low priority for now.

**Comment by deekoder** [comment]
Can this be closed then?

**Comment by vadmeste** [comment]
> Can this be closed then?

Yes.
